feat: added total time for outage in notification msg

prabhsharan36 opened the pull request:
# Description
Resolves #177

## Type of change
I have changed the msg body which goes to notification providers to give the total time for the outage.

Please delete any options that are not relevant.

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings

## Screenshots

![Screenshot 2023-07-10 020006](https://github.com/louislam/uptime-kuma/assets/56472269/c489e3a5-a78d-460d-9bfa-67f078069c73)

CommanderStorm commented on server/model/monitor.js:
```diff
@@ -809,6 +809,24 @@ class Monitor extends BeanModel {
+                                0
+                            ]);
+                            const downTime = dayjs(mostRecentDownBeat?.time).diff(dayjs(downAfterPreviousUpBeat?.time), "minutes", true);
+                            bean.msg = `${bean.msg} <Down for ${Math.floor(downTime)} mintue(s) ${(downTime % 1).toFixed(2).replace(/^0\./, "")} second(s)>`;
```
Adding a message to the `msg` is something which might not be ideal.
This prevents notification providers like `discord`/`slack` to including them in the fancy embeds.

This is something that Luis will have to give his opinion on (both paths are workable imo)

chakflying replied:
I think a better solution would be to add a parameter `downDuration` to send(), which is the duration of downtime in seconds, then individual notification providers can format the message as needed. This would provide better customization, and be non-breaking for the existing notifications.

prabhsharan36 replied:
@chakflying I guess we will send the same msg to every notification provider. For instance, we will send how many minutes or hours this service goes down in the same manner to every provider. In refactored code, I have put the syntax like (mm: ss) format.

chakflying replied:
If you check provider `discord.js`, it doesn't even use the `msg` parameter for up/down notifications. To be clear, I'm not saying this PR needs to update all the notification providers. You can update the ones that you are using. I'm just saying this is not a good solution considering the current situation.

prabhsharan36 replied:
> If you check provider discord.js, it doesn't even use the msg parameter for up/down notifications.

Oh, I now get your point. Thanks for letting me know. I will add one more parameter in the send() function which will have downDuration value in seconds and then all notification providers can use it the way they want.

CommanderStorm replied:
I think this change should be added to the `heartbeatJSON` instead. 
What is the rationale behind adding another parameter?

If adding it to `heartbeatJSON` is not an option / adding a parameter this is better, please add this parameter to every function invocation to make this obvious to the authors of the notificationproviders)

prabhsharan36 replied:
I have added downTime to hearbeatJSON. 
![code1](https://github.com/louislam/uptime-kuma/assets/56472269/f7bacba3-592e-440c-903b-30bdf62722d8)


But in the "sendNotification" function I have kept the downTime parameter and not added another property to the "bean" parameter as eventually we are storing the data of "bean" into the database and adding a property to it will mean either we have to add coulmn in table which is of no use or we have to delete that added (downTime) property from the "bean" object.

![code](https://github.com/louislam/uptime-kuma/assets/56472269/09ecf323-1232-47f2-854a-f313f40b5076)

CommanderStorm reviewed (changes requested):
I have made some style and logic comments which were unclear to me.
Given that this targets all notification providers, I tried to be thorough.

prabhsharan36 commented on server/model/monitor.js:
```diff
@@ -809,6 +809,24 @@ class Monitor extends BeanModel {
+                    // Change the notification msg if status becomes UP from DOWN
+                    // Adding msg for the downtime
+                    if (bean.status === UP) {
+                        const downAfterPreviousUpBeat = await R.findOne("heartbeat", " monitor_id = ? AND status = ? AND time > ( SELECT time from heartbeat WHERE monitor_id = ? AND status = ? ORDER BY time DESC ) ORDER BY time ASC", [
```
> Do we need to do special accounting for MAINTANANCE or PENDING?

According to me we shouldn't count them into the downtime period as in MAINTANANCE the service may be down but that is expected and in PENDING status we don't know whether service is UP or DOWN.

> What happens in this case? (is this reported as down 0 minute(s) 0 second(s)?)

I have changed this condition now and now I am adding the duration of down beats.

> Would the current beat's UP to the last UP (or first event) be more accurate?

According to me, that should be the condition, but I am happy about your views also on this.

prabhsharan36 commented on server/model/monitor.js:
```diff
@@ -809,6 +809,24 @@ class Monitor extends BeanModel {
+                                0
+                            ]);
+                            const downTime = dayjs(mostRecentDownBeat?.time).diff(dayjs(downAfterPreviousUpBeat?.time), "minutes", true);
+                            bean.msg = `${bean.msg} <Down for ${Math.floor(downTime)} mintue(s) ${(downTime % 1).toFixed(2).replace(/^0\./, "")} second(s)>`;
```
![image](https://github.com/louislam/uptime-kuma/assets/56472269/9e5aeae5-d814-47ed-a609-6ea795255304)

I have changed the format to this. But, I am not very satisfied with the minutes keyword at last as in this case we only have seconds and 00 minutes. Could you suggest a better format for this?

prabhsharan36 commented on server/model/monitor.js:
```diff
@@ -809,6 +809,24 @@ class Monitor extends BeanModel {
+                    // Change the notification msg if status becomes UP from DOWN
+                    // Adding msg for the downtime
+                    if (bean.status === UP) {
+                        const downAfterPreviousUpBeat = await R.findOne("heartbeat", " monitor_id = ? AND status = ? AND time > ( SELECT time from heartbeat WHERE monitor_id = ? AND status = ? ORDER BY time DESC ) ORDER BY time ASC", [
```
> Given the current scaling issues, this is not ideal, but given that this query is only executed on notification if the bean is UP, this probably won't be an issue.

Yes, basically this query will only run when the service goes from DOWN to UP which is not a daily scenario for most of the service. Also, for cutting down the rows I am using filtering to get only those related to specific monitor by it's ID.
There is also no other way we can get the most recent UP beat without using the ordering of rows.

> Have you measured the impact of this query? How long does it take? (sqlite does not allow concurrent queries => long-running queries could cause downtime on bigger databases or slower IO (e.g. rasperry Pis using SD-Cards))
Could waiting here for longer be an issue?

This actually varies from system to system and also how big the table size is. In my system, there are only around 100 rows for a single monitor and it gives results in some milliseconds from DB. So, we can't really find out how it will impact other systems of different sizes.

RobertCSternberg commented:
Many SMS services will deny sending messages containing special characters to the carrier networks.  It would make the notification more friendly to read and possibly pass through a few more notification services if we removed the encapsulating <> characters.

prabhsharan36 commented:
> Many SMS services will deny sending messages containing special characters to the carrier networks. It would make the notification more friendly to read and possibly pass through a few more notification services if we removed the encapsulating <> characters.

I have already changed this and now the response looks something like this:
![Screenshot 2023-07-12 015020](https://github.com/louislam/uptime-kuma/assets/56472269/de9c57fe-4548-4e2e-b3e0-b7c504ec23d3)

prabhsharan36 commented on server/model/monitor.js:
```diff
@@ -809,6 +809,24 @@ class Monitor extends BeanModel {
+                    if (bean.status === UP) {
+                        const downAfterPreviousUpBeat = await R.findOne("heartbeat", " monitor_id = ? AND status = ? AND time > ( SELECT time from heartbeat WHERE monitor_id = ? AND status = ? ORDER BY time DESC ) ORDER BY time ASC", [
+                            this.id,
+                            0,
```
it's done

prabhsharan36 commented on server/model/monitor.js:
```diff
@@ -809,6 +809,24 @@ class Monitor extends BeanModel {
+                                0
+                            ]);
+                            const downTime = dayjs(mostRecentDownBeat?.time).diff(dayjs(downAfterPreviousUpBeat?.time), "minutes", true);
+                            bean.msg = `${bean.msg} <Down for ${Math.floor(downTime)} mintue(s) ${(downTime % 1).toFixed(2).replace(/^0\./, "")} second(s)>`;
```
it's done

prabhsharan36 commented on server/model/monitor.js:
```diff
@@ -809,6 +809,24 @@ class Monitor extends BeanModel {
+                                this.id,
+                                0
+                            ]);
+                            const downTime = dayjs(mostRecentDownBeat?.time).diff(dayjs(downAfterPreviousUpBeat?.time), "minutes", true);
```
it's done

prabhsharan36 commented on server/model/monitor.js:
```diff
@@ -809,6 +809,24 @@ class Monitor extends BeanModel {
+                                0
+                            ]);
+                            const downTime = dayjs(mostRecentDownBeat?.time).diff(dayjs(downAfterPreviousUpBeat?.time), "minutes", true);
+                            bean.msg = `${bean.msg} <Down for ${Math.floor(downTime)} mintue(s) ${(downTime % 1).toFixed(2).replace(/^0\./, "")} second(s)>`;
```
I guess the before one looks more readable

CommanderStorm commented:
Could you add `Resolves #3301` to the description? (this way github will close that issue automatically)

CommanderStorm commented:
Sorry to bother you again, but I found more issues ^^
Could you also add 
 
`Resolves #177`
`Resolves #1215`
`Resolves #428`

CommanderStorm commented on server/notification.js:
```diff
@@ -135,9 +135,9 @@ class Notification {
      * @throws Error with fail msg
      */
-    static async send(notification, msg, monitorJSON = null, heartbeatJSON = null) {
+    static async send(notification, msg, monitorJSON = null, heartbeatJSON = null, downTime = null) {
```
I think this change should be added to the `heartbeatJSON` as well.
What is the rationale behind adding another parameter?

If adding it to `heartbeatJSON` is not an option / adding a parameter this is better, please add this parameter to every function invocation to make this obvious to the authors of the notificationproviders)

CommanderStorm commented:
You know the drill by now (sigh):
`Resolves #2362`
`Resolves #1158`

CommanderStorm commented on server/model/monitor.js:
```diff
@@ -809,8 +809,39 @@ class Monitor extends BeanModel {
                 if (Monitor.isImportantForNotification(isFirstBeat, previousBeat?.status, bean.status)) {
+                    // Change the notification msg if status becomes UP from DOWN
+                    // Adding msg for the downtime
+                    let downTimeInSeconds;
```
```suggestion
                    let downTimeInSeconds = null;
```

Computroniks replied:
Why bother setting it to `null`? It is already `undefined` so it doesn't really make much of a difference.

CommanderStorm commented on server/model/monitor.js:
```diff
@@ -809,8 +809,39 @@ class Monitor extends BeanModel {
+                    }
                     log.debug("monitor", `[${this.name}] sendNotification`);
-                    await Monitor.sendNotification(isFirstBeat, this, bean);
+                    let downTime = 0;
+                    if (downTimeInSeconds) {
+                        downTime = downTimeInSeconds.duration;
+                    }
+                    await Monitor.sendNotification(isFirstBeat, this, bean, downTime);
```
```suggestion
                    await Monitor.sendNotification(isFirstBeat, this, bean, downTimeInSeconds.duration);
```

prabhsharan36 replied:
What if downTimeInSeconds is undefined here, isn't downTimeInSeconds.duration will give error?

CommanderStorm commented on server/model/monitor.js:
```diff
@@ -1233,7 +1264,7 @@ class Monitor extends BeanModel {
      * @param {Bean} bean Status information about monitor
      */
-    static async sendNotification(isFirstBeat, monitor, bean) {
+    static async sendNotification(isFirstBeat, monitor, bean, downTime = 0) {
```
```suggestion
    static async sendNotification(isFirstBeat, monitor, bean, downTime = null) {
```

Computroniks replied:
I am not sure I really understand the point of setting it to `null`

Computroniks replied:
I am not sure I really understand the point of setting it to null

CommanderStorm commented on server/model/monitor.js:
```diff
@@ -1259,6 +1290,7 @@ class Monitor extends BeanModel {
                     heartbeatJSON["timezone"] = await UptimeKumaServer.getInstance().getTimezone();
                     heartbeatJSON["timezoneOffset"] = UptimeKumaServer.getInstance().getTimezoneOffset();
                     heartbeatJSON["localDateTime"] = dayjs.utc(heartbeatJSON["time"]).tz(heartbeatJSON["timezone"]).format(SQL_DATETIME_FORMAT);
+                    heartbeatJSON["downTime"] = downTime;
```
```suggestion
                    if (downTime !== null) {
                        heartbeatJSON["downTime"] = downTime;
                    }
```

Computroniks replied:
Would a simple truthiness check not be fine here?

chakflying replied:
Being able to distinguish between information not available and "0" is a good thing here IMO.

prabhsharan36 replied:
Sending the downTime property in heartbeatJSON even when downTime is 0 is a good thing as it tells the notification handler that there is a property called  "downTime" in heartbeatJSON which can be anywhere between 0 - [MAX NUMBER]. Giving the downTime property only when it is larger than 0 will confuse the notification handler where the property sometimes comes and sometimes does not.

CommanderStorm reviewed (changes requested):
The current implementation does add the `downTime` parameter, even if this does not make sense
⇒ The people who will use this parameter will use it wrong.

I have attached an alternative way of handling this

Computroniks commented on server/notification-providers/telegram.js:
```diff
@@ -9,9 +10,17 @@ class Telegram extends NotificationProvider {
         let okMsg = "Sent Successfully.";
 
         try {
+            let message = msg;
```
Not sure what the point of this is?

prabhsharan36 replied:
This is added as it is recommended not to change the function attributes directly, so I am saving into another variable and changing it further by concatenating it with another string.

Computroniks commented:
Oops, sorry about the duplication. GitHub has for some reason added my replies also as comments.

louislam commented:
Unfortunately, as I am changing the uptime calculation method, I believe it is not working in 2.0.0 anymore.

Since the message is removed, be sure that you have read this and better have a discussion first next time.
> ⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

mabed-fr commented:
> Malheureusement, comme je change la méthode de calcul de la disponibilité, je pense qu'elle ne fonctionne plus dans la version 2.0.0.
> 
> Modifié dans la version 2.0.0 :
> 
> * La table des battements de cœur stocke uniquement les données sur 24 heures
> * heartbeat.duration n'est plus utilisé
> * Les données seront regroupées dans des tableaux `stat_daily`et`stat_minutely`
> 
> Qui s'approprie ce pr.
> 
> Puisque le message a été supprimé, assurez-vous d'avoir lu ceci et feriez mieux d'avoir une discussion la prochaine fois.
> 
> > ⚠️⚠️⚠️Puisque nous n'acceptons pas tous les types de demandes de tirage et que nous ne voulons pas vous faire perdre du temps. Assurez-vous d'avoir lu les règles des pull request : 
> > [https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime -kuma](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma)

Do you have an idea to still enable this feature?

CommanderStorm commented:
@prabhsharan36 
Thank you for this contribution. Unfortunately, in `v2.0` we changed the way how we calculate uptimes a lot. This makes this change incompatible as stated by Louis in https://github.com/louislam/uptime-kuma/pull/3394#issuecomment-1725356957.

Given how different this approach is from what is needed and the progress this has made in the last few months, I am going to close this issue.

As documented in our [contribution guide](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md), we believe that new features create less wasted effort if a discussion takes place beforehand (⇒ we can tell you if we are about to make significant changes to this area of code).
This feature is still possible and should likely be implemented via f.ex. adding a field such as `last_up_heartbeat_at` on each monitor. If the necessary changes are made, we can reopen the PR or handle this in a different PR ^^
